src/dmd/expressionsem.d

```diff
@@ -10177,6 +10177,129 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+                }
+                arguments.push(exp.e1);
+                arguments.push(exp.e2);
+
+                exp.isCTFEHookPart = true;
+                Expression ce = new CallExp(exp.loc, id, arguments).expressionSemantic(sc);
+
+                *output = ce.expressionSemantic(sc);
+            }
+            else if (exp.op == EXP.concatenateElemAssign)
+            {
+                if (auto ve = exp.e1.isVarExp)
+                {
+                    import core.stdc.ctype : isdigit;
+                    // The name of the indices array that static foreach loops uses.
+                    // See dmd.cond.lowerNonArrayAggregate
+                    enum varName = "__res";
+                    const(char)[] id = ve.var.ident.toString;
+                    if (id.length > varName.length && id[0 .. varName.length] == varName && id[varName.length].isdigit)
+                        return;
+                }
+
+                Identifier hook = global.params.tracegc ? Id._d_arrayappendcTXTrace : Id._d_arrayappendcTX;
+                if (!verifyHookExist(exp.loc, *sc, Id._d_arrayappendcTXImpl, "appending element to arrays", Id.object))
+                    return setError();
+
+                // Lower to object._d_arrayappendcTXImpl!(typeof(e1))._d_arrayappendcTX{,Trace}(e1, 1), e1[$-1]=e2
+                Expression id = new IdentifierExp(exp.loc, Id.empty);
+                id = new DotIdExp(exp.loc, id, Id.object);
+                auto tiargs = new Objects();
+                tiargs.push(exp.e1.type);
+                id = new DotTemplateInstanceExp(exp.loc, id, Id._d_arrayappendcTXImpl, tiargs);
+                id = new DotIdExp(exp.loc, id, hook);
+                id = id.expressionSemantic(sc);
+
+                auto arguments = new Expressions();
+                arguments.reserve(5);
+                if (global.params.tracegc)
+                {
+                    auto funcname = (sc.callsc && sc.callsc.func) ? sc.callsc.func.toPrettyChars() : sc.func.toPrettyChars();
+                    arguments.push(new StringExp(exp.loc, exp.loc.filename.toDString()));
+                    arguments.push(new IntegerExp(exp.loc, exp.loc.linnum, Type.tint32));
+                    arguments.push(new StringExp(exp.loc, funcname.toDString()));
+                }
+
+                Expression eValue1;
+                Expression value1 = extractSideEffect(sc, "__appendtmp", eValue1, exp.e1);
+
+                arguments.push(value1);
+                arguments.push(new IntegerExp(exp.loc, 1, Type.tsize_t));
+
+                Expression ce = new CallExp(exp.loc, id, arguments);
+                ce = ce.expressionSemantic(sc);
+
+                Expression eValue2;
+                Expression value2 = exp.e2;
+                if (!value2.isVarExp && !value2.isIntegerExp &&
+                    !value2.isStructLiteralExp && !value2.isArrayLiteralExp &&
+                    !value2.isAssocArrayLiteralExp)
```

Comment:
What is this checking for?

Reply:
When lowering an expression like `a ~= a[$ - 1]`, `$` is evaluated after calling `_d_arrayappendcTX`. The hook extends the array, thus changing its length and therefore leading to an incorrect evaluation of `$`.

This code forces every non-trivial expression to be saved to a temporary variable. Simply calling `extractSideEffect` would not work because it does not create temporaries for `IndexExp`s.

Reply:
The corresponding check in the original code:
```D
if (e2.Eoper != OPvar && e2.Eoper != OPconst)
                {
                    // Evaluate e2 and assign result to temporary s2.
                    // Do this because of:
                    //    a ~= a[$-1]
                    // because $ changes its value
```
Why is there no comment anymore, and why does it include struct/array literals? I think that goes wrong when doing something like `a ~= [ a[$-1] ];`.

Reply:
I found a better approach: I imitated the previous code from e2ir.d here.